### PR TITLE
[MLIR][omp] Add omp.workshare op

### PR DESCRIPTION
**Pull request by ivanradanov**
1/4 in stack for workshare implementation

1/4 https://github.com/llvm/llvm-project/pull/101443
2/4 https://github.com/llvm/llvm-project/pull/101444
3/4 https://github.com/llvm/llvm-project/pull/101445
4/4 https://github.com/llvm/llvm-project/pull/101446

### CODE REVIEW

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp [hunk 1]
```diff
@@ -1683,6 +1683,19 @@ LogicalResult SingleOp::verify() {
...
+// WorkshareOp
+//===----------------------------------------------------------------------===//
+
+void WorkshareOp::build(OpBuilder &builder, OperationState &state,
+                        const WorkshareOperands &clauses) {
+  WorkshareOp::build(builder, state, clauses.nowait);
+}
+
+LogicalResult WorkshareOp::verify() {
+  return (*this)->getRegion(0).getBlocks().size() == 1 ? success() : failure();
```
Comment (tblah) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:1696]: Please could you add an error message if this check fails and a test in `mlir/test/Dialect/OpenMP/invalid.mlir`.

**Review by tblah — COMMENTED** [review]
Thanks for your work on this. Please could you also add a parse/print test in `mlir/test/Dialect/OpenMP/ops.mlir`

**Review by kiranchandramohan — COMMENTED** [review]
Please add some roundtrip tests in mlir/test/Dialect/OpenMP/ops.mlir.

The questions here would be:
1. Should we have an alternative representation like `hlfir.elemental` that exposes the implicit loop of the constructs in it.
2. Should enforce additional constraints on the operations inside the body of the workshare construct? Like should they have loop interfaces?

**Review by tblah — APPROVED** [review]
LGTM. Thanks for the updates

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td [hunk 2]
```diff
@@ -287,6 +287,49 @@ def SingleOp : OpenMP_Op<"single", traits = [
...
+    block before the work gets divided between the threads. See the flang
+    LowerWorkshare pass for details.
+  }] # clausesDescription;
+
+  let builders = [
+    OpBuilder<(ins CArg<"const WorkshareOperands &">:$clauses)>
+  ];
+}
+
+def WorkshareLoopWrapperOp : OpenMP_Op<"workshare_loop_wrapper", traits = [
```
Comment (skatrak) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:316]: Nit: According to the discussion in [this RFC](https://discourse.llvm.org/t/rfc-uniformize-openmp-dialect-operation-names/77715), I think something along these lines should be preferred:
```suggestion
def WorkshareLoopWrapperOp : OpenMP_Op<"workshare.loop_wrapper", traits = [
```

**Review by skatrak — APPROVED** [review]
Thank you, this LGTM as well. Just a small nit about the name for `WorkshareLoopWrapperOp`.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td (hunk 2 above)
Comment (ivanradanov) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td]: Thank you for the suggestion, done

**Pull request merged**